tool/planet/agent.go

```diff
@@ -250,23 +249,46 @@ func updateDNS(conf *LeaderConfig, hostname string, newMasterIP string) error {
 	return nil
 }
 
-var electedUnits = []string{
+var controlPlaneUnits = []string{
 	"kube-controller-manager.service",
 	"kube-scheduler.service",
 	"kube-apiserver.service",
 }
 
-func unitsCommand(command string) error {
-	log.Debugf("executing %v on %v", command, electedUnits)
+func startUnits(ctx context.Context) error {
+	log.Debug("Start control plane units.")
 	var errors []error
-	for _, unit := range electedUnits {
-		cmd := exec.Command("/bin/systemctl", command, unit)
-		log.Debugf("executing %v", cmd)
-		out, err := cmd.CombinedOutput()
+	for _, unit := range controlPlaneUnits {
+		logger := log.WithField("unit", unit)
+		err := systemctlCmd(ctx, "start", unit)
 		if err != nil {
 			errors = append(errors, err)
 			// Instead of failing immediately, complete start of other units
-			log.Warningf("failed to execute %v: %s\n%v", cmd, out, trace.DebugReport(err))
+			logger.WithError(err).Warn("Failed to start unit.")
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+func stopUnits(ctx context.Context) error {
+	log.Debug("Stop control plane units.")
+	var errors []error
+	for _, unit := range controlPlaneUnits {
+		logger := log.WithField("unit", unit)
+		err := systemctlCmd(ctx, "stop", unit)
+		if err != nil {
+			errors = append(errors, err)
+			// Instead of failing immediately, complete stop of other units
+			logger.WithError(err).Warn("Failed to stop unit.")
+		}
+		// Even if 'systemctl stop' did not fail, the service could have failed stopping
+		// even though 'stop' is blocking, it does not return an error upon service failing.
+		// See github.com/gravitational/gravity/issues/1209 for more details
+		if err := systemctlCmd(ctx, "is-failed", unit); err == nil {
```

Comment:
Sorry, I just realized something I missed on the non-backport PR. Since we pass no-block to the underlying systemctl command, will a unit enter the failed state before we check the is-failed status, since presumably the unit is allowed some amount of time to shutdown?

Reply:
We don't pass `no-block` to `systemctl` here - only etcd-related functionality does - so it should be good given the possibility that dbus command could still fail which we'll address with reconciler loop changes.

Reply:
Ah my bad, I was looking at the wrong function signature when that stood out to me.